File: 5. Longest Palindromic Substring.py
# https://leetcode.com/problems/longest-palindromic-substring/

# Given a string s, return the longest palindromic substring in s.

################################################################################

# palindrome -> dp
# dp[i][j] = s[i...j] is good
# dp[i][j] = s[i+1...j-1] is good and s[i] == s[j]
# time: O(n^2), space: O(n^2)

class Solution:
    def longestPalindrome(self, s: str) -> str:
        if not s: return None
        n = len(s)
        if n == 1: return s
        
        max_len_start = max_len_end = 0
        dp = [[False] * n for _ in range(n)]
        
        # base case
        for i in range(n): # single char str
            dp[i][i] = True
        for i in range(n - 1): # double char str
            if s[i] == s[i + 1]:
                dp[i][i + 1] = True
                max_len_start, max_len_end = i, i + 1
        
        # loop for incr
        for incr in range(2, n): # incr = 2, ..., n - 1
            for i in range(n - incr): # i = 0, ..., n - 1 - incr
                j = i + incr # s[i...j]
                if dp[i + 1][j - 1] and s[i] == s[j]:
                    dp[i][j] = True
                    max_len_start, max_len_end = i, j

        return s[max_len_start : max_len_end + 1]

# palindrome -> expand around center
# expand around 'a' and 'ab' for every char in string
# time: O(n^2), space: O(1)

# class Solution:
    def expand_center(self, s):
        if not s: return None
        n = len(s)
        if len(s) == 1: return s
        
        max_len = 0
        max_len_start = max_len_end = 0
        
        for i in range(n):
            len_a, start_a, end_a = self.get_len(i, i, s) # expand around 'a'
            len_ab, start_ab, end_ab = self.get_len(i, i + 1, s) # expand around 'ab'
            
            if len_a > max_len:
                max_len = len_a
                max_len_start, max_len_end = start_a, end_a
                
            if len_ab > max_len:
                max_len = len_ab
                max_len_start, max_len_end = start_ab, end_ab

        return s[max_len_start : max_len_end + 1]
    
    # Expand around center
    def get_len(self, l, r, s):
        if l < 0 or r >= len(s): return 0, None, None
        
        while l >= 0 and r < len(s) and s[l] == s[r]:
            l -= 1
            r += 1
        return r - l - 1, l + 1, r - 1 # length, start_idx, end_idx

File: 5. Longest Palindromic Substring.cpp
// https://leetcode.com/problems/longest-palindromic-substring/

// Given a string s, return the longest palindromic substring in s.

////////////////////////////////////////////////////////////////////////////////

// palindromic -> dp[i][j] = if s[i...j] is pali
class Solution {
public:
    string longestPalindrome(string s) {
        int n = s.size();
        vector<vector<bool>> dp(n, vector<bool>(n, false));
        int maxLen = 1, start = 0;
        // len = 1
        for (int i = 0; i < n; ++i) dp[i][i] = true;
        // len = 2
        for (int i = 0; i + 1 < n; ++i) {
            if (s[i] == s[i+1]) {
                dp[i][i+1] = true;
                maxLen = 2; start = i;
            }
        }
        // len = 3...n
        for (int len = 3; len <= n; ++len){
            // check s[i...i+len-1]
            for (int i = 0, j = i + len - 1; j < n; ++i, ++j) {
                if (s[i] == s[j] and dp[i+1][j-1]) {
                    dp[i][j] = true;
                    maxLen = len; start = i;
                }
            }
        }
        return s.substr(start, maxLen);
    }
};
